File: connection-examples/java/subscribe.java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.util.Properties;
import java.sql.ResultSet;
import java.sql.Statement;

public class App {

    private final String url = "jdbc:postgresql://MATERIALIZE_HOST:6875/materialize";
    private final String user = "MATERIALIZE_USERNAME";
    private final String password = "MATERIALIZE_PASSWORD";

    /**
     * Connect to Materialize
     *
     * @return a Connection object
     */
    public Connection connect() throws SQLException {
        Properties props = new Properties();
        props.setProperty("user", user);
        props.setProperty("password", password);
        props.setProperty("ssl","true");

        return DriverManager.getConnection(url, props);

    }

    public void subscribe() {
        try (Connection conn = connect()) {

            Statement stmt = conn.createStatement();
            stmt.execute("BEGIN");
            stmt.execute("DECLARE c CURSOR FOR SUBSCRIBE my_view");
            while (true) {
                ResultSet rs = stmt.executeQuery("FETCH ALL c");
                if(rs.next()) {
                    System.out.println(rs.getString(1) + " " + rs.getString(2) + " " + rs.getString(3));
                }
            }
        } catch (SQLException ex) {
            System.out.println(ex.getMessage());
        }
    }

    public static void main(String[] args) {
        App app = new App();
        app.subscribe();
    }
}

File: connection-examples/typescript/src/subscribe.ts
// node 14+ cjs named exports not found in pg
import pkg from 'pg';
import State, { Update } from './state';
const { Client } = pkg;

const client = new Client({
  user: "MATERIALIZE_USERNAME",
  database: "materialize",
  password: "APP_SPECIFIC_PASSWORD",
  host: "MATERIALIZE_HOST",
  port: 6875,
  ssl: true
});

interface CounterSum {
  sum: number;
}

async function main() {
  try {
    await client.connect();

    await client.query('BEGIN');
    await client.query('DECLARE c CURSOR FOR SUBSCRIBE (SELECT sum FROM counter_sum) WITH (PROGRESS);');

    const state = new State<CounterSum>();
    const buffer: Array<Update<CounterSum>> = [];

    // Loop indefinitely
    while (true) {
      const { rows } = await client.query('FETCH ALL c');
      rows.forEach(row => {
        // Map row fields
        const {
          mz_timestamp: ts,
          mz_progressed: progress,
          mz_diff: diff,
          sum,
         } = row;

        //  When a progress is detected, get the last values
        if (progress) {
          if (buffer.length > 0) {
            try {
              state.update(buffer, ts);
            } catch (err) {
              console.error(err);
            } finally {
              buffer.splice(0, buffer.length);
            }
          }
        } else {
            buffer.push({ value: { sum }, diff });
        }
      });
    }
  } catch (e) {
    console.log("error: ", e);
  } finally {
    client.end(); // only on error
  }
}

await main();